### PR TITLE
Core/Maps: move pooling hand-off outside of Map::CheckRespawn

### LINKED ISSUE #25777: Core/Maps: BaseMapLoadAllGrids causes server crash
**Description:**
When starting the server with `GridUnload = 0` and` BaseMapLoadAllGrids = 1` , the server crashes while preloading the maps.

**Expected behaviour:**
The server should not crash.

**Steps to reproduce the problem:**
apply the following settings in `world.conf` and start the server:
```
#
#    GridUnload
#        Description: Unload grids to save memory. Can be disabled if enough memory is available
#                     to speed up moving players to new grids.
#        Default:     1 - (enable, Unload grids)
#                     0 - (disable, Do not unload grids)

GridUnload = 0

#
#    BaseMapLoadAllGrids
#        Description: Load all grids for base maps upon load. Requires GridUnload to be 0.
#                     This will take around 5GB of ram upon server load, and will take some time
#                     to initially load the server.
#        Default:     0 - (Don't preload all base maps, dynamically load as used)
#                     1 - (Preload all grids in all base maps upon load)

BaseMapLoadAllGrids = 1
```

The following exception is thrown, thanks to @illfated for catching with debug build:
```
Exception code: C0000420 An assertion failure has occurred.

Assertion message:
C:\Trinity\Source\src\server\game\Maps\Map.h:854 in Map::GetRespawnMapForType ABORTED.
```
https://gist.github.com/illfated/46a28ab26a4428e3e268e59ccebb35d6

(linked issue: https://github.com/TrinityCore/TrinityCore/issues/25409)

**Branch(es):**
3.3.5

**TC rev. hash/commit:**
TrinityCore rev. 3b1e911 2020-09-06 01:40:00 +0200 (3.3.5 branch) (Win64, Release, Static)

**Operating system:**
Win10

### PR DESCRIPTION
**Pull request by Treeston**
This ensures that pooling logic can't stomp all over our nice Map internals

fixes #25777

### CODE REVIEW

File: src/server/game/Maps/Map.cpp [hunk 1]
```diff
@@ -3038,45 +3038,41 @@ bool Map::CheckRespawn(RespawnInfo* info)
 
     uint32 poolId = info->spawnId ? sPoolMgr->IsPartOfAPool(info->type, info->spawnId) : 0;
     // Next, check if there's already an instance of this object that would block the respawn
-    // Only do this for unpooled spawns
```
Comment (Treeston) [inline src/server/game/Maps/Map.cpp:3041]: this entire diff block is just un-indenting an `if`

File: src/server/game/Maps/Map.cpp [hunk 2]
```diff
@@ -3094,21 +3090,6 @@ bool Map::CheckRespawn(RespawnInfo* info)
         info->respawnTime = respawnTime;
         return false;
     }
-
```
Comment (Treeston) [inline src/server/game/Maps/Map.cpp:3097]: this is an actual removal

File: src/server/game/Maps/Map.cpp [hunk 3]
```diff
@@ -3269,33 +3250,67 @@ void Map::DoRespawn(SpawnObjectType type, ObjectGuid::LowType spawnId, uint32 gr
 
 void Map::ProcessRespawns()
 {
+    // defer pool hand-off until after the loop to ensure internal consistency before going to external logic
```
Comment (Treeston) [inline src/server/game/Maps/Map.cpp:3253]: new stuff 1/3

File: src/server/game/Maps/Map.cpp [hunk 4]
```diff
@@ -3269,33 +3250,67 @@ void Map::DoRespawn(SpawnObjectType type, ObjectGuid::LowType spawnId, uint32 gr
...
+    std::vector<PoolQueueEntry> poolQueue;
+
     time_t now = GameTime::GetGameTime();
     while (!_respawnTimes.empty())
     {
         RespawnInfo* next = _respawnTimes.top();
         if (now < next->respawnTime) // done for this tick
             break;
+
+        if (uint32 poolId = sPoolMgr->IsPartOfAPool(next->type, next->spawnId)) // is this part of a pool?
```
Comment (Treeston) [inline src/server/game/Maps/Map.cpp:3270]: new stuff 2/3

File: src/server/game/Maps/Map.cpp [hunk 5]
```diff
@@ -3269,33 +3250,67 @@ void Map::DoRespawn(SpawnObjectType type, ObjectGuid::LowType spawnId, uint32 gr
...
-        {
+        else
+        { // new respawn time, update heap position
             ASSERT(now < next->respawnTime); // infinite loop guard
             _respawnTimes.decrease(next->handle);
             SaveRespawnInfoDB(*next);
         }
     }
+
+    // ok, now that every is safely stashed away again, tell pooling system to do its thing
```
Comment (Treeston) [inline src/server/game/Maps/Map.cpp:3299]: new stuff 3/3

File: src/server/game/Maps/Map.cpp (hunk 1 above)
Comment (jackpoz) [reply src/server/game/Maps/Map.cpp:3041]: image

this setting makes the diff more readable

Comment (Treeston) [reply src/server/game/Maps/Map.cpp:3041]: nice, didn't know that

Comment (illfated) [reply src/server/game/Maps/Map.cpp:3041]: Also easily available by just slapping `?w=1` at the end of the URL:
- https://github.com/TrinityCore/TrinityCore/pull/25785/files?w=1

**Pull request merged**